limbo/slack.py

```diff
@@ -139,40 +111,79 @@ def connect_slack_websocket(self, ws_url):
         except:
             raise SlackConnectionError
 
-    def parse_channel_data(self, channel_data):
-        for channel in channel_data:
-            if "name" not in channel:
-                channel["name"] = channel["id"]
-            if "members" not in channel:
-                channel["members"] = []
+    def _dig(self, obj, *keys):
+        for key in keys:
+            if not obj or key not in obj:
+                return None
+            obj = obj[key]
+        return obj
 
-            self.attach_channel(channel['name'], channel['id'], channel['members'])
+    def get_all(self, api_method, collection_name):
+        """
+        Return all objects in an api_method and handle pagination.
+
+        For example, "users.list" returns an object like:
 
-    def parse_user_data(self, user_data):
-        for user in user_data:
+        {
+            "members": [{<member_obj>}, {<member_obj_2>}],
+            "response_metadata": {
+                "next_cursor": "cursor_id"
+            }
+        }
+
+        so if you call `get_all("users.list", "members")`, this function
+        will return all member objects to you while handling pagination
+        """
+        objs = []
+        # if you don't provide a limit, the slack API won't return a cursor to you
+        page = json.loads(self.api_call(api_method, limit=25))
+        while 1:
+            for obj in page[collection_name]:
+                objs.append(obj)
+
+            cursor = self._dig(page, "response_metadata", "next_cursor")
+            if cursor:
+                page = json.loads(self.api_call(api_method, cursor=cursor))
+            else:
+                break
+
+        return objs
+
+    def get_channel_list(self):
+        # this call may or may not provide members for each channel, so
+        # let's not rely on the members being in it. If we need them
+        # (which I don't think we do?) we can get them later
```

Comment:
the slack docs suggest that the default channels.list call will get `members`. from your comment, i'm guessing that you weren't seeing that data. weird!

we _could_ reduce the size of this call by including `exclude_members` = True since we won't be using that data. that would involve adding an extra param onto `get_all()`, which will make that function more confusing. i'm fine with just... not doing that

Reply:
you're right, I misread the docs to mean that they would be excluding members automatically. Their section on `exclude_members` and `limit` is very confusing. They say:

> exclude_members may not work well with pagination at this time. Channels with very large memberships and teams with many channels may cause the method to throw a HTTP 500 error. Please exclude memberships with exclude_members

Which doesn't make sense because they tell you that `exclude_members` may not work well (in what way?) and then suggest that you use it.

`get_all` should probably just take a `**kwargs` argument to pass on to the method being called, I'll add that anyway to ease any case like this in the future.